### DISCUSSION TITLE
索引是否生效问题

### DISCUSSION THREAD

**Discussion by hanqiushi** [milvus-io/milvus#19142] [Q&A and General discussion] [answered]
您好，有几个问题想请教下
1. 比如建立hnsw类型索引，从哪里能看出来建立的索引是否生效了
2. 执行search的时候能看出来是否走索引了
3. 从log看默认16384条数据会作为一个bulk进行插入，如果这时milvus宕机了，数据会丢失么

**Comment by yhmo** [comment, marked as answer]
1. By the index_building_progress()接口 you can know how many rows have been indexed. The following links are pymilvus&java SDK documentation for this interface:
https://milvus.io/api-reference/pymilvus/v2.1.1/Utility/index_building_progress().md
https://milvus.io/api-reference/java/v2.1.0/Index/getIndexBuildProgress().md

2. search()的时候，对于 indexed segment，走index search；对于 non-indexed segment，走brute-force search
3. insert()的处理流程：proxy node先接收到insert request，对传过来的data按primary key进行hash，根据hash值以及shard number值分成多份message，然后把这些message写入pulsar，然后datanode从pulsar订阅消息，拿到消息后把消息中包含的数据写入minio。而insert()接口在数据写入pulsar后立即返回，由于这些数据已经被持久化在pulsar中，如果这时宕机，datanode会重新从pulsar中拉取数据，数据不会丢失。如果insert()接口在数据还未写入pulsar就宕机，那么客户端就没有接收到insert成功的返回信息，因此客户端可以知道该数据没有插入成功，可在服务重启后重新插入，也就没有数据丢失的说法。
